src/observer.ts

```diff
@@ -12,10 +12,29 @@ export function observer<P extends object, TRef = {}>(
 ): React.MemoExoticComponent<
     React.ForwardRefExoticComponent<React.PropsWithoutRef<P> & React.RefAttributes<TRef>>
 >
+
 export function observer<P extends object>(
     baseComponent: React.FunctionComponent<P>,
     options?: IObserverOptions
 ): React.FunctionComponent<P>
+
+export function observer<
+    C extends React.FunctionComponent<any> | React.RefForwardingComponent<any>,
+    Options extends IObserverOptions
+>(
+    baseComponent: C,
+    options?: Options
+): Options extends { forwardRef: true }
+    ? C extends React.RefForwardingComponent<infer TRef, infer P>
+        ? C &
+              React.MemoExoticComponent<
+                  React.ForwardRefExoticComponent<
+                      React.PropsWithoutRef<P> & React.RefAttributes<TRef>
+                  >
+              >
+        : never /* forwardRef set for a non forwarding component */
+    : C & React.FunctionComponent
```

Comment:
Does `React.FunctionComponent` only declare the static members, or also the function taking props itself? As in the last case it would result in an intersection of `fn(props: P) & fn(props: any)` which is effectively `fn(props: any)`? If you need to pass in the generic `P`, you might need an additional conditional type and `infer P` on `C`

Reply:
If I'm confusing or you're stuck, let me know and I'll take a look myself. If the type tests are completely to uncover the original issue, it should be fine.

Reply:
It effectively declares only static members because the default is object `FunctionComponent<P = {}>`, so it doesn't get intersected with `any`, but passing actual props would work as well.